src/config.ts

```diff
@@ -0,0 +1,84 @@
+import {cosmiconfigSync} from 'cosmiconfig';
+import {Validator} from 'jsonschema';
+import {merge} from 'lodash';
+
+export interface SnowpackConfig {
+  options: {
+    babel?: boolean;
+    clean?: boolean;
+    dest?: string;
+    exclude?: string[];
+    externalPackage?: string[];
+    include?: string[];
+    nomodule?: string;
+    nomoduleOutput?: string;
+    noSourceMap?: boolean;
+    optimize?: boolean;
+    remotePackage?: string[];
+    remoteUrl?: string;
+    strict?: boolean;
+  };
+  webDependencies: string[];
+}
+
+const defaultConfig: SnowpackConfig = {
+  options: {
+    clean: false,
+    dest: 'web_modules',
+    exclude: ['**/__tests__/*', '**/*.@(spec|test).@(js|mjs)'],
+    noSourceMap: false,
+    nomoduleOutput: 'app.nomodule.js',
+    optimize: false,
+    remoteUrl: 'https://cdn.pika.dev',
+    strict: false,
+  },
+  webDependencies: [],
+};
+
+const schema = {
+  id: '/config',
+  type: 'object',
+  properties: {
+    options: {
+      type: 'object',
+      properties: {
+        babel: {type: 'boolean'},
+        clean: {type: 'boolean'},
+        dest: {type: 'string'},
+        exclude: {type: 'array', items: {type: 'string'}},
+        externalPackage: {type: 'array', items: {type: 'string'}},
+        include: {type: 'array', items: {type: 'string'}},
+        nomodule: {type: 'string'},
+        nomoduleOutput: {type: 'string'},
+        noSourceMap: {type: 'boolean'},
+        optimize: {type: 'boolean'},
+        remotePackage: {type: 'array', items: {type: 'string'}},
+        remoteUrl: {type: 'string'},
+        strict: {type: 'boolean'},
+      },
+      required: [],
+    },
+    webDependencies: {type: 'array', items: {type: 'string'}},
+  },
+};
+
+export default async function loadConfig(cliFlags?: SnowpackConfig) {
+  const explorerSync = cosmiconfigSync('snowpack');
+  const result = explorerSync.search(); // search for snowpack config
+  const config: SnowpackConfig = result.config;
+
+  // user has no config
+  if (!config || result.isEmpty) {
+    return cliFlags ? merge(defaultConfig, cliFlags) : defaultConfig;
+  }
+
+  // validate against schema; throw helpful user if invalid
+  const {validate} = new Validator();
+  validate(config, schema, {throwError: true});
+
+  // if valid, apply config over defaults
+  const mergedConfig = merge(defaultConfig, config);
+
+  // return CLI flags if present; otherwise return config
+  return cliFlags ? merge(mergedConfig, cliFlags) : mergedConfig;
+}
```

Comment:
So here’s my thinking so far. I‘ll organize my thinking more into “subjects” than steps.

**cosmiconfig**
Great little project! This basically searches `package.json`, then `.snowpackrc`, then `.snowpack.config.js`, etc. It’ll also parse the file. But the only downside is it **won’t validate.** How can we be friendly to users, to let them know whether or not the config is valid?

**jsonschema**
This is something I picked up from webpack’s config—it’s a library that validates a schema against yours. I initially thought about just manually doing a lot of `if / then` logic to make sure certain keys were there, but even for the small config we have now, that’s a lot of code! And it’s a lot of manual entry.

I didn’t initially start out thinking schema validation was important, but now I do. I think showing users helpful feedback on what’s wrong immediately is a better user experience. And jsonschema is a standard way, and the least amount of code I could think to use to achieve this.

**Merging**
The order of merging matters. This is the order in which we load (later order means it overrides the previous):

1. Default config
2. cosmiconfig config file (`package.json`, `.snowpackrc`, `.snowpack.config.js`, …)
3. CLI flags

Does all this make sense? If this is tracking, then I can start manually testing this, and adding tests both to ensure the config works, and that helpful errors are thrown.

Reply:
+100, that all sounds exactly right. I'd recommend pushing the JSON schema work off into a second PR since this one is already getting pretty big, but if you're jamming on it all together already then I'm still happy to review at whatever size :)

Reply:
Ah sorry 😅. If it’s all right, I’d like to keep it. I added a test to (hopefully) deal with the overhead of reviewing.

Reply:
np then, happy to review :)